### ISSUE TITLE
Badarg error in crypto Erlang on Kubernetes (OCP) with FIPS enabled

### ISSUE THREAD

**Issue by dtma007** [erlang/otp#4818]
I'm not certain if this is a bug or a configuration issue, but hoping to get some guidance, if possible.

I have a RabbitMQ/Erlang cluster deployed on Kubernetes (OCP 4.7).  On non-FIPS systems, our RabbitMQ deployment works fine.  But on FIPS systems, we are seeing RabbitMQ/Erlang crash with error:

```
bad argument in call to crypto:evp_generate_key_nif(x25519, undefined) in ssl_cipher:generate_client_shares/2 line 1390
```

For context, here are some of the lines in the log:
```
[debug] <0.271.0> GET https://kubernetes.default.svc.cluster.local:443/api/v1/namespaces/xxx/endpoints/xxx-rabbitmq-ha-discovery

[error] <0.448.0> CRASH REPORT Process <0.448.0> with 1 neighbours crashed with reason: bad argument in call to crypto:evp_generate_key_nif(x25519, undefined) in ssl_cipher:generate_client_shares/2 line 1390

[debug] <0.271.0> Response: {error,{failed_connect,[{to_address,{"kubernetes.default.svc.cluster.local",443}},{inet,[inet],{eoptions,{{badarg,[{crypto,evp_generate_key_nif,[x25519,undefined],[]},{ssl_cipher,generate_client_shares,2,[{file,"ssl_cipher.erl"},{line,1390}]},{tls_connection,init,3,[{file,"tls_connection.erl"},{line,590}]},{gen_statem,loop_state_callback,11,[{file,"gen_statem.erl"},{line,1166}]},{tls_connection,init,1,[{file,"tls_connection.erl"},{line,153}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]},{gen_statem,call,[<0.448.0>,{start,2250},infinity]}}}}]}}
```

In our "advanced.config" we have tried to enable FIPS with this setting:
```
[

    {crypto, [{fips_mode, true}]}

].
```

Our env has:
- Erlang 23.3.2 
- RabbitMQ Server 3.8.15 
- OpenSSL 1.1.1g-15.el8_3
- the container is RHEL 8.3

from: https://github.com/rabbitmq

Is there something I might be missing/misconfigured?  Thanks!

**Comment by HansN** [comment]
x25519 is not available in FIPS mode.  The availability of algorithms could be checked dynamically with crypto:supports/0 or crypto:supports(curves) for e.g. x25519 .

You have OpenSSL 1.1.1g.  As far as I know OpenSSL cryptolib is not FIPS validated nor available in cryptolib versions above 1.0.2.

Check what crypto:info_fips() returns: not_supported, not_enabled or enabled.

**Comment by HansN** [comment]
Hi @dtma007 ,
First a disclaimer: I know nothing about RabbitMQ.

If you need FIPS mode, OpenSSL libcrypto 1.0.2 is the latest that is certified for FIPS. 1.0.2 is however not supported for open-source any longer by OpenSSL and receives no security updates.  There is a special procedure to compile it as FIPS compliant.  The Erlang/OTP then need to be configured with the `--enable-fips` configure flag before compiling and then call `crypto:enable_fips_mode(true)`. This should probably be done in RabbitMQ if they support the FIPS mode, otherwise by the top application.

OpenSSL 3.0 will have a FIPS module they say, but 3.0 is not supported by OTP yet.  We have it on the story list, but it is not prioritized and I can't promise when it will be.

If you need x25519 you must have 1.1.1 but it has no FIPS support (or the by us yet unsupported 3.0 probably in non-FIPS mode).

I'll check with our ssl expert about the details of the error message you provided.

**Comment by dtma007** [comment]
@HansN Thanks for your reply.   I don't think we specifically require x25519, but I'm not sure how to exclude it from being used.  I have been experimenting with the `ssl_options` parameters in configuration file to specify specific ciphers.  Do you know which ciphers should be excluded in order to avoid using `x25519`?

Here are the settings I'm trying to use, but it hasn't worked so far.  Thanks

```
    ssl_options.cacertfile            = /etc/certs/rabbitmq.pem
    ssl_options.certfile              = /etc/certs/rabbitmq.crt
    ssl_options.keyfile               = /etc/certs/rabbitmq.key
    ssl_options.verify                = verify_none
    ssl_options.fail_if_no_peer_cert  = false
    ssl_options.versions.1            = tlsv1.3
    ssl_options.versions.2            = tlsv1.2
    ssl_options.ciphers.1  = ECDHE-ECDSA-AES256-GCM-SHA384
    ssl_options.ciphers.2  = ECDHE-RSA-AES256-GCM-SHA384
    ssl_options.ciphers.3  = ECDHE-ECDSA-AES256-SHA384
    ssl_options.ciphers.4  = ECDHE-RSA-AES256-SHA384
    ssl_options.ciphers.5  = ECDH-ECDSA-AES256-GCM-SHA384
    ssl_options.ciphers.6  = ECDH-RSA-AES256-GCM-SHA384
    ssl_options.ciphers.7  = ECDH-ECDSA-AES256-SHA384
    ssl_options.ciphers.8  = ECDH-RSA-AES256-SHA384
    ssl_options.ciphers.9  = DHE-RSA-AES256-GCM-SHA384
    ssl_options.ciphers.10 = DHE-DSS-AES256-GCM-SHA384
    ssl_options.ciphers.11 = DHE-RSA-AES256-SHA256
    ssl_options.ciphers.12 = DHE-DSS-AES256-SHA256
    ssl_options.ciphers.13 = ECDHE-ECDSA-AES128-GCM-SHA256
    ssl_options.ciphers.14 = ECDHE-RSA-AES128-GCM-SHA256
    ssl_options.ciphers.15 = ECDHE-ECDSA-AES128-SHA256
    ssl_options.ciphers.16 = ECDHE-RSA-AES128-SHA256
    ssl_options.ciphers.17 = ECDH-ECDSA-AES128-GCM-SHA256
    ssl_options.ciphers.18 = ECDH-RSA-AES128-GCM-SHA256
    ssl_options.ciphers.19 = ECDH-ECDSA-AES128-SHA256
    ssl_options.ciphers.20 = ECDH-RSA-AES128-SHA256
    ssl_options.ciphers.21 = DHE-RSA-AES128-GCM-SHA256
    ssl_options.ciphers.22 = DHE-DSS-AES128-GCM-SHA256
    ssl_options.ciphers.23 = DHE-RSA-AES128-SHA256
    ssl_options.ciphers.24 = DHE-DSS-AES128-SHA256
    ssl_options.ciphers.25 = ECDHE-ECDSA-AES256-SHA
    ssl_options.ciphers.26 = ECDHE-RSA-AES256-SHA
    ssl_options.ciphers.27 = DHE-RSA-AES256-SHA
    ssl_options.ciphers.28 = DHE-DSS-AES256-SHA
    ssl_options.ciphers.29 = ECDH-ECDSA-AES256-SHA
    ssl_options.ciphers.30 = ECDH-RSA-AES256-SHA
    ssl_options.ciphers.31 = ECDHE-ECDSA-AES128-SHA
    ssl_options.ciphers.32 = ECDHE-RSA-AES128-SHA
    ssl_options.ciphers.33 = DHE-RSA-AES128-SHA
    ssl_options.ciphers.34 = DHE-DSS-AES128-SHA
    ssl_options.ciphers.35 = ECDH-ECDSA-AES128-SHA
    ssl_options.ciphers.36 = ECDH-RSA-AES128-SHA
```

**Comment by IngelaAndin** [comment]
You should be configuring  supported groups!

Default but no support for  Edwards curves (x25519 and x488) (Will be used with ECDHE cipher suites)
```

 {supported_groups,  [secp256r1,
                                  secp384r1]}
 
```
or all ECC (except Edwards) and DH groups (will be used with DHE cipher suites)

```
 {supported_groups,   [secp256r1,
                                    secp384r1,
                                    secp521r1,
                                    ffdhe2048,
                                    ffdhe3072,
                                    ffdhe4096,
                                    ffdhe6144,
                                    ffdhe8192]}
```

**Comment by dtma007** [comment]
@IngelaAndin Hi Ingela,
Thanks for your reply.
If I specify the supported_groups, would I still still need to call out specific ciphers?  Or should the "supported_groups" be sufficient?

Also, do you happen to know where the "supported_groups" would be specified relative to the other SSL settings?

For example, in RabbitMQ config, the settings seem to be given like this:

**Comment by IngelaAndin** [comment]
The supported groups options goes together with all the other options in the options list. 

In TLS-1.3 key exchange algorithm is not part of the cipher suite, all of the cipher suites can be combined with
any key exchange algorithms defined by supported groups. Note that TLS-1.2 and TLS-1.3 has no overlapping cipher suites.

These are the TLS-1.3 cipher suites
```
["TLS_AES_256_GCM_SHA384","TLS_AES_128_GCM_SHA256",
 "TLS_CHACHA20_POLY1305_SHA256","TLS_AES_128_CCM_SHA256",
 "TLS_AES_128_CCM_8_SHA256"]
```

**Comment by michaelklishin** [comment]
RabbitMQ does not call the FIPS mode function and does not expose supported groups in its ini-style (Cuttlefish) config format. Both would be easy to add but without FIPS mode support in OpenSSL 1.1, which all modern Erlang packages usually target, I don't see much point in doing that. Once OpenSSL 1.1 supports FIPS, let us know :)

**Comment by michaelklishin** [comment]
TLSv1.3 tends to be an all-or-nothing proposition but can be a decent alternative to FIPS **in terms of ticking security architect boxes**. RabbitMQ TLS guide has a [dedicated section on it](https://www.rabbitmq.com/ssl.html#tls1.3). I'd be curious to know if TLSv1.3 only setups would be considered "good enough" by those looking for FIPS.

**Comment by dtma007** [comment]
@michaelklishin Thanks for your help.   From your understanding, would the only way to make RabbitMQ work in a FIPS-enabled container as of today now would be with OpenSSL 1.0.2 and older versions of Erlang and RabbitMQ?   I noticed this post from about 2 years on the rabbitmq-users forum:  https://groups.google.com/g/rabbitmq-users/c/uk9hcAcMNUA?pli=1

**Comment by HansN** [comment]
A for the same - or a similar - problem is included in the release OTP-24.1.,
